### ISSUE TITLE
Refactor linear problem dynamics-related code in ReachContinuous/ReachDiscrete classes and provide a dense test coverage

### ISSUE THREAD

**Issue by pgagarinov** [SystemAnalysisDpt-CMC-MSU/ellipsoids#37]
Deadline - end of the semester
----------------------------------------

1) Create an abstract protected method in AReach called "getProbDynamicsBuilder" that accepts two logica inputs: isDisturbance and isBackward and returns
a problem definition constructor. The Constructor should be returned as function handles, for instance for
ReachDiscrete, isBackward=true, isDisturbance=false the output would be 
@(varargin)gras.ellapx.lreachplain.probdyn.LReachDiscrBackwardDynamics(gras.ellapx.lreachplain.probdef.LReachContProblemDef(varargin{:})) and for ReachContinous for the
same type of system - @gras.ellapx.lreachuncert.probdyn.LReachProblemDynamicsFactory.createByParams. 
This is because these calls correspond to the existing logic in ReachDiscrete.getProbDynamics and elltool.reach.ReachContinuous.getProbDynamics. 
Once getProbDynamicsBuilder is implemented in both ReachContinous and ReachDiscrete merge the implementation of getProbDynamics methods from ReachContinous
and ReachDiscrete by placing its code in AReach and making it call getProbDynamicsBuilder. This way the code getProbDynamics will be common for both ReachContinous
and ReachDiscrete while getProbDynamicsBuilder will contain all the difference between continuous and discrete systems.

2) Now, once getProbDynamics is in AReach replace calcPrecision input argument with two separate arguments - relTol and absTol, make sure that whereever
getProbDynamics is calls both relTol and absTol are passed as inputs. Right now only relTol is passed - taken 
from self.relTol field, you should also pass self.absTol. Please note that getProbDynamicsBuilder should return such builders for Discrete systems that 
can accept both relTol and absTol even for Discrete systems. However for Discrete systems relTol and absTol should not be passed any further (i.e. ignored).
Here is an example of ignoring input argument in an anonymous function that should help to understand:
fBuild=@(varargin)constructDiscrete(varargin{1:end-2}); In this example last two arguments (relTol and absTol) are ignored. Thus, to keep things simple it is better
to make relTol and absTol last input arguments in builders returned by getProbDynamicsBuilder. For continuous systems all the classes that previously 
accepted calcPrecision should now accept relTol and absTol.  One of this classes is 
+gras\+ellapx\+lreachplain\+probdyn\LReachProblemDynamicsInterp.m (line 30).

        function odePropList=getOdePropList(self,calcPrecision)
            odePropList={'NormControl',self.ODE_NORM_CONTROL,'RelTol',...
                calcPrecision*self.CALC_PRECISION_FACTOR,...
                'AbsTol',calcPrecision*self.CALC_PRECISION_FACTOR};
        end
		
defined in C:\Users\Administrator\Documents\_Git\ellipsoids\products\+gras\+ellapx\+lreachplain\+probdyn\AReachProblemDynamics.m
should be changed so that it looks like this:

        function odePropList=getOdePropList(self,relTol, absTol)
            odePropList={'NormControl',self.ODE_NORM_CONTROL,'RelTol',...
                relTol*self.REL_TOL_FACTOR,...
                'absTol',absTol*self.ABS_TOL_FACTOR};
        end

where REL_TOL_FACTOR and ABS_TOL_FACTOR constant have the same value as CALC_PRECISION_FACTOR for now. CALC_PRECISION_FACTOR constant need to be removed.

3) Add one more optional input to gras\+ellapx\+lreachuncert\+probdyn\LReachProblemDynamicsFactory.createByParams called "isDiscrete", if not specified assume 
that isDiscrete=false. Then make createByParams a bit smarter by implementing an ability to construct discrete systems in the same way
as getSysWithDisturb and getSysWithoutDisturb did that in ReachDiscrete that prior to implementation of step 1) above.
Then get rid of getProblemDynamics call in both AReach constructor and evolveApprox and use 
gras\+ellapx\+lreachuncert\+probdyn\LReachProblemDynamicsFactory.createByParams instead. Please note that there is no need to pass isDisturbance into createByParams
as it is capable to determining if the system contains disturbance in the same way as AReach can. This means that isDisturbance method in AReach
should be removed as well as the following call on line 564:

            isDisturbance = LReachProblemDynamicsFactory.getIsDisturbance(...
                ctStrCMat, qtStrCMat);

4) Cover problem dynamics and problem definition classes from the following packages with tests 
\products\+gras\+ellapx\+lreachplain\+probdef
\products\+gras\+ellapx\+lreachplain\+probdyn
\products\+gras\+ellapx\+lreachuncert\+probdef
\products\+gras\+ellapx\+lreachuncert\+probdyn

the test cases should be located in the following packages

\products\+gras\+ellapx\+lreachplain\+probdef\+test\+mlunit
\products\+gras\+ellapx\+lreachplain\+probdyn\+test\+mlunit
\products\+gras\+ellapx\+lreachuncert\+probdef\+test\+mlunit
\products\+gras\+ellapx\+lreachuncert\+probdyn\+test\+mlunit

functions run_tests should be located in each +test subpackage.
and their calls should be added to \products\+gras\+ellapx\+test\run_tests function.

For the problem definition classes (in probdef packages) the tests should construct the problem definition objects on a few simple systems 
and verify that all methods return a correct result. Please note that there should be no copy-pasting in the tests - you should parameterize them 
with tested calsses so that the same test case can be applied to different problem definition classes. This approach is very well demonstrated
in elltool.reach.test.run_cont_tests - please use this function as a reference.

For problem dynamics classes the approach should be similar - create problem dynamics for a few simple systems and make sure that the values returned by 
its methods (by getAtDynamics method responsible for returning a matrix function for A(t)) match the expected result on a certain time grid. For A(t) it means that 
you need to check that A(t_i) matches the expected result for all t_i from the time grid.

After each step (and actually during the implementation of the steps) you need to make sure that ALL tests pass. Implementation without all the tests passing = zero result.

**Comment by ilya2015** [comment]
After analyzing the task i have few questions:
1) You sad: "After each step you need to make sure that ALL tests pass". Do I understand correctly that there are not any tests for this steps?
2) Where i can describe work of problem definition constructor?

**Comment by pgagarinov** [comment]
1) ReachContinuous and ReachDiscrete are covered with tests located in elltool.reach.test package. For continuous systems you run run_cont_tests and for discrete - run_discr_tests

2) Not sure I understand the question. Do you want to read about the constructor? Well, the best way is to read source code and study the real-life usage examples (in ReachContinuous/ReachDiscrete classes as well as in gras.ellapx.uncertcalc.run function.

**Comment by ilya2015** [comment]
Yesterday you sad, that i can create getProbDynamics function instead her Abstract version in AReach. But ReachDiscrete and ReachContinuous according to different use timevec. Maybe i must use getProbDynamicsBuilder insted getSysWithDisturb and getSysWithoutDisturb in ReachDiscrete and ReachContinuous classes?

**Comment by pgagarinov** [comment]
No, you need to proceed as we discussed and I do not understand why it is a problem that timeVec is different. 

First of all, it is an object-oriented programming, not functional programming. There are no functions here, getProbDynamics  is a class method, not a function. One of the basic principles of OOP is inheritance which allows to re-use code of base classes. getProbDynamics  is going to be implemented in AReach and used in both of its subclasses. getProbDynamics  will call an abstract method getProbDynamicsBuilder that is going to be defined differently in different subclasses. Why does it matter if timeVec is different, I cannot understand.

**Comment by ilya2015** [comment]
I understood about re-use code. But now i have small problem.. Matlab doesn't see getProbDynamicsBuilder from getProbDynamics (i tried this in AReach and ReachContinuous, describing Builder before our function). Also i can't run elltool.reach.ReachContinuous. ... getProbDynamicsBuilder  becouse it is Abstract protected method.

**Comment by pgagarinov** [comment]
I belive the problem is that you do not have any experience with OOP in Matlab. I suggest you create a few simple classes - BaseClass.m and SubClass1.m and SubClass2.m where both SubClass1 and SubClass2 inherits BaseClass. The in BaseClass create an abstract methodX and public methodY that calls method X. Then in SubClass1 and SubClass2 provide different implementations of these methods and make sure that everything works. Once you do that I'm sure you will understand what is wrong with your real-life implementation. For more information about Matlab OOP please refer to http://www.mathworks.com/help/matlab/object-oriented-programming.html

**Comment by ilya2015** [comment]
I read Help about OOP. In task there is phrase: "an abstract protected method in AReach called "getProbDynamicsBuilder" ". But if i want to call it from Static method getProbDynamics, this Builder must be a Static too?

**Comment by pgagarinov** [comment]
getProbDynamics is not static.

**Comment by ilya2015** [comment]
![image](https://cloud.githubusercontent.com/assets/15003196/11453131/0cb5d24c-9610-11e5-874c-4041da3fc981.png)
Sorry, maybe we tell about different getProbDynamics ?

**Comment by pgagarinov** [comment]
Ah, yes, you are right. Let us make getProbDynamics not-static, just protected. Then it will be able to call non-static methods.

**Comment by ilya2015** [comment]
I did the first part of task (as i it understood). Tests for AReach passed.

**Comment by pgagarinov** [comment]
So far so good but step 1 is not complete because getProbDynamics still needs to be moved to AReach (by merging code from both ReachContinuous and ReachDiscrete).

**Comment by ilya2015** [comment]
Did you mean this merge?

**Comment by pgagarinov** [comment]
You moved getDynamics to AReach - that is good but what you did after is wrong. One of basic principles of OOP is polymorphism which in our context means that getProbDynamics cannot know anything about its subclasses. I.e. subclasses (ReachContinuous and ReachDiscrete know about AReach but not vice versa). Thus you need to write getProbDynamics in a generic way such that it works regardless of class. This means that you cannot use isa to check a class of self. But I don't think you even need to - the logic is generic and should not depend on whether it is a discrete system or continuous.

**Comment by ilya2015** [comment]
I understand that my merge doesn't differ of previous variant. But i don't know what we do with "Calcprecision": it is needed in ReachContinious, but it doesn't used in Discrete case (last parameter from call handleObj). There is the same situation with timevec.

**Comment by ilya2015** [comment]
I hope that i finished first part.

**Comment by pgagarinov** [comment]
I agree that 1) is done but please add 1-2 lines of comments to the place where you use "varargin{1:end-1}" that a potential reader of your code understand why, particularly explain which argument you want to ignore. And please check your English grammar in those comments. Once this is done you need to merge into master and re-create you branch.

**Comment by pgagarinov** [comment]
Please also remove round brackets in if conditions like in "if (~isDiscrete)" and the merge to master using our merge policy.

**Comment by ilya2015** [comment]
In your example:
```
   function odePropList=getOdePropList(self,relTol, absTol)
        odePropList={'NormControl',self.ODE_NORM_CONTROL,'RelTol',...
            relTol*self.REL_TOL_FACTOR,...
            'absTol',calcPrecision*self.ABS_TOL_FACTOR};
    end
```
instead calcPrecision you mean absTol?
P.S. Your previous note i will do in new branch. Can i delete branch Issue_37_IliaSh?

**Comment by pgagarinov** [comment]
Yes, I fixed the description. Yes, you can delete your branch now.

**Comment by ilya2015** [comment]
Probably i did second part of task.

**Comment by pgagarinov** [comment]
So, is step 2 or not? Also, please eliminate all spaces you inserted on step1.

**Comment by ilya2015** [comment]
I wrote previous comment before Jenkins sent me results. I checked only reach tests, but there are many errors in another places.
Sure, i will remove spaces.

**Comment by ilya2015** [comment]
I think that i finished second part.

**Comment by pgagarinov** [comment]
3) Please check all classes that you changed and make sure that MLint gives no warnings (box in the right upper corner should be green). 
4) Please apply an automatic source code indentation (Ctrl+A, Ctrl+I).
5) Please explain why do we have to use [min(varargin{end-2}) max(varargin{end-2})] for continuous systems and just varargin{end-2} for discrete systems? Also, even if there is a valid reason for doing that we need to make sure that probDefConstr receives timeLimVec instead of timeVec for continuous systems. This can be done by changing those places that call probDefConstr returned by getProbDynamicsBuilder. The idea is that the interface of the constructor is the same for both continuous and discrete systems.
6) Answering both 1) and 2) above: the idea is that all problem dynamics construction logic should be moved FROM AReach/ReachContinuous/Discrete into LReachProblemDynamicsFactory. Then all places where you called getProblemDynamics you should now call createParams. We do that for one reason - to have one central place where all dynamics construction logic is located. Then we will be able to use this logic from different places with AReach/ReachContinuous/Discrete classes being only one of them, among other modules that will benefit from using the same logic is command-line function gras.ellapx.uncertcalc.run that calculates ellipsoidal approximations for continuous systems.

**Comment by ilya2015** [comment]
3) I can not delete some warnings in ReachContinuous (line 181) and ReachDiscrete (line 247) - i must pass parameter 'self'. Also Can i suppress they?
5) We don't use isBackword for Continuous systems, and we convert timevec to [min(timevec) max(timevec)]. Now in AReach.getProbDynamics i create handleObj which can take different timevector. So i don't know how i can do the same interface for for both continuous and discrete systems.

**Comment by pgagarinov** [comment]
3) I do not see why you can't replace self with ~ - you do not seem to use self. ~ means that you just do not use the corresponding input argument, it doesn't mean that you do not expect it to be passed.

5) What you do with isBackward is just fine.

Can you point me to a specific place in ReachContinuous that passes in timeVec, not timeLimVec?

>So i don't know how i can do the same interface for both continuous and discrete systems.
This can be done by changing ReachContinuous so that it passes in timeLimVec instead of timeVec. You just need to find a specific place where this happens.

**Comment by ilya2015** [comment]
Can i merge to master current branch? (actual versions of first and second parts)

**Comment by pgagarinov** [comment]
Before the merge please do the following

6) Please fix the way MLint warnings are removed in the places listed below. Instead of using "~" just use "Suppress warning on this line". This is because by using ~ you make it very difficult to recall that ~ designated "absTol" or "relTol". 

classdef AReachProblemLTIDynamics<...
        gras.ellapx.lreachplain.probdyn.AReachProblemDynamics
    methods
        function self=AReachProblemLTIDynamics(problemDef,relTol,absTol)


classdef AReachProblemDynamicsInterp<...
        gras.ellapx.lreachplain.probdyn.AReachProblemDynamics
    methods
        function self=AReachProblemDynamicsInterp(problemDef,relTol,absTol)


        function probDefConstr = getProbDynamicsBuilder (self, ...
                isDisturbance,~)

7) Can you please add a simple test to the new test case (that you need to create)
elltool.reach.test.mlunit.ContinuousMiscSimpleTC

that verifies that using [min(timeVec) max(timeVec)] doesn't work. This can be a very simple test for a very simple system, even 2-dimensional, it just has to fail when you use min(timeVec),max(timeVec) and should pass when you use sort(timeVec). And please add this test case to elltool.reach.test.run_cont_tests.

**Comment by ilya2015** [comment]
I read the help for \products\+elltool\+reach\ReachContinuous.m (line 252) and there is next description:
```
timeVec: double[1, 2] - time interval
```
This mean that author understand that we always give to constructor vector with only two elements. Probably one my change with getProbDynamicsBuilder can not change this logic.

**Comment by pgagarinov** [comment]
Ok, this make sense, can you then please rename all such entries of timeVec to timeLimVec. Just make sure that this is the same timeVec. Also please replace min/max in

        function [ellTubeRel,goodDirSetObj] = auxMakeEllTubeRel(self, probDynObj, ...
                l0Mat, timeVec, isDisturb,absTol,relTol,approxTypeVec)
            import gras.ellapx.enums.EApproxType;
            import gras.ellapx.lreachplain.GoodDirsContinuousFactory;
            import modgen.common.throwerror;
            %
            timeVec = [min(timeVec) max(timeVec)];

with sort.

**Comment by ilya2015** [comment]
I am sorry, but are you sure that we need number of timevec's elements > 2? Becouse for continious systems timevec means interval of time and it consist of two elements. If you sure then i must do many changes in places which is used for continuous systems.

**Comment by pgagarinov** [comment]
I think there is a misunderstanding here. First of all - I agree with you that we only need 2 elements. Secondly I asked you to make just a variable rename - from timeVec (that in my mind corresponds to an array of times like [0, 0.1, 0.2, ... 11].) to timeLimVec that in my mind corresponds to a to element vector of time limits. 

The reason I ask you to do this rename is that in EllTube classes and in many other places timeVec variable names corresponds to an array of time moments, not to a time limits. So I do want to have a naming consistency so that when we see timeVec we know that it is an array of length>2 while timeLimVec always means that it is a two element vector of time limits.

Does it make sense?

**Comment by ilya2015** [comment]
Oh, i understand you. In function getProbDynamics we use timeLimVec for continious systems. 
Of course i can rename this vector in other places. Also you told about creating ContinuousMiscSimpleTC, but this check (format of timevec) there is already in AReach.

**Comment by pgagarinov** [comment]
timeLimVec doesn't have to be renamed if it corresponds to 2 element vector, It is timeVec that needs to be renamed to timeLimVec if it does correspond to 2 element vector. What I do not want is you renaming timeVec to timeLimVec where it is not necessary so please be very careful, you cannot just automatically rename all timeVec entries into timeLimeVec.

When I suggested you implement a test in ContinuousMiscSimpleTC I didn't realize that timeVec in ReachContinuous' constructor is just a two element vector. I think this request should be dropped for now so just forget about it.

**Comment by ilya2015** [comment]
I have one question again: you told that i should move all problem dynamics construction logic to LReachProblemDynamicsFactory. But getProbDynamics is called only from AReach. How can i understand value of 'isDiscrete' ? I think that check with operator 'isa' is a bad idea.

**Comment by pgagarinov** [comment]
In AReach

     isDisturbance = LReachProblemDynamicsFactory.getIsDisturbance(...
                ctStrCMat, qtStrCMat);

so you need to move this call into LReachProblemDynamicsFactory so that the factory class determines a type of the system all by itself.

**Comment by ilya2015** [comment]
I remember about this call, but we use 'isDisturbance' in both systems. And if i have this value then i don't know type of our system.

**Comment by pgagarinov** [comment]
Yes, I agree that we won't be able to call createByParams from AReach anymore because we need to know a value of isDiscrete parameter that is passed into createByParams. So let us just call it from ReachContinuous and ReachDiscrete. Since it will be just a single call of createByParams there won't be any copy-pasting, do you agree?

**Comment by ilya2015** [comment]
Yes, if we called createByParams  from ReachContinuous/ReachDiscrete then we have not any problem. But i don't understand how one call can help us.. Maybe if i add static field to LReachProblemDynamicsFactory then i can understand whitch is the type of system.

**Comment by pgagarinov** [comment]
Forget about AReach, you remove all traces of getProbDynamics  from AReach and add direct calls of LReachProblemDynamicsFactory.createByParams   from  ReachContinuous and ReachDiscrete. 

When I was talking about one call I mean that there won't be any copy pasting between ReachContinous and ReachDiscrete because from these classes you will just call createByParams, all the logic will be inside createByParams. See?

**Comment by ilya2015** [comment]
Yes, i know that i will move logic getProbDynamics and getProbDynamicsBuilder to createByParams.
I mean next: for example now we are calling getProbDynamics from AReach's constructor. And i should remove this call and add call createByParams to ReachContinuous and ReachDiscrete. Maybe we can change part of parent's constructor from them?

**Comment by pgagarinov** [comment]
What do you mean by "Maybe we can change part of parent's constructor from them?"? Please be more descriptive.

But anyways, here is out goal - to move all the construction logic to the factory. You said - we do not know if the system is discrete if we call this factory from AReach. I replied - look, we do not have to call it from AReach, we can call it directly from ReachContinuous/ReachDiscrete classes. 

If you can the factory from the derivative classes there is no need for any logic located in AReach - you can call createByParams directly from RCont or RDiscr and pass in either isDiscrete=false or true. 

Now, you might ask - why in hell did we bother with moving a common code to AReach if we are going to remove everything from AReach anyway on step 3). For two reasons:

a) To make you understand how all this stuff works and how inheritance works in Matlab
b) To prepare a solid ground for moving the code into the factory. This thing is that when you move the construction logic into the factory you will still need to avoid the copy pasting that you removed on steps 1 and 2. Thus you will have to put the code from AReach into a separate private method inside the factory that will be called just like AReach.getProbDynamics. 

Makes sense?

**Comment by ilya2015** [comment]
Sorry, I mean that i did not understand - where are the places in ReachContinuous or ReachDiscrete for call createByParams, so now we are not using getProbDynamics in these classes explicitly.

**Comment by pgagarinov** [comment]
a) we remove getProbDynamics from AReach and define it as abstract in AReach
b) define getProbDynamics such that it does nothing but calls createByParam

You might think - why did I move getProbDynamics from ReachContinuous and Discrete to AReach if I now need to move it back? That is because current body getProbDynamics will go into a private method of the factory. 

step 3.1 current body of AReach.getProbDynamics -> private method of the factory
step 3.2 remove getProbDynamics from AReach
step 3.3 declare getProbDynamics as abstract in AReach
step 3.4 implement getProbDynamics in RCont and RDiscr such that it calls createByParams.

Makes sense?

**Comment by ilya2015** [comment]
Yes, thank you very much.

**Comment by pgagarinov** [comment]
Before you start doing 3) let us merge the refined steps 1) and 2) please.

**Comment by ilya2015** [comment]
I did third part. Tests for Reach passed well, but there is one failure in tests for lib. Until I realized how connected lib and reach (tracebacks from log are not very usefull).

**Comment by pgagarinov** [comment]
Yes, you can.

**Comment by ilya2015** [comment]
It is strange, but all tests passed. Can you check third part please?

**Comment by pgagarinov** [comment]
Well, not quite what I expected to see.

AReach had the following code:
        function linSys = getProbDynamics(self,atStrCMat,btStrCMat, ...
                ptStrCMat,ptStrCVec,ctStrCMat,qtStrCMat,qtStrCVec, ...
                x0Mat,x0Vec,timeLimVec,relTol,absTol,isDisturb)
            isBack = timeLimVec(1) > timeLimVec(2);
            handleObj = self.getProbDynamicsBuilder(isDisturb,isBack);
            if isDisturb
                linSys = handleObj(atStrCMat,btStrCMat,ptStrCMat, ...
                    ptStrCVec,ctStrCMat,qtStrCMat,qtStrCVec,x0Mat, ...
                    x0Vec,timeLimVec,relTol,absTol);
            elseif ~isDisturb
                linSys = handleObj(atStrCMat,btStrCMat,ptStrCMat, ...
                    ptStrCVec,x0Mat,x0Vec,timeLimVec,relTol,absTol);
            end
        end

I expected to see this in a private method of the factory.

Instead I see probDefConstr = getProbDynDiscrBuild(isDisturb,isBack) which has Discr in its name?
The whole point of having the private method is to have the logic that is common for both continuous and discrete systems in separate place.

**Comment by ilya2015** [comment]
Did you mean this structure? But i needed a new function getSysWithDisturb which means old version of createByParams.

**Comment by pgagarinov** [comment]
8) There is a cyclic reference in

        function probDefConstr = getContProbDynamicsBuilder(isDisturb,isBack) %#ok<INUSD>
            %
            % input argument varargin{end - 2} is a 'timeLimVec'; we
            % transform it to sort(varargin{end-2}) for Continuous systems
            %
            if ~isDisturb
                probDefConstr = @(varargin)gras.ellapx.lreachplain. ...
                    probdyn.LReachProblemDynamicsFactory. ...
                    createByParams (varargin{1:end-3}, ...
                    sort(varargin{end-2}),varargin{end-1:end});
            elseif isDisturb
                probDefConstr = @(varargin)gras.ellapx.lreachuncert. ...
                    probdyn.LReachProblemDynamicsFactory.getSysWithDisturb( ...
                    varargin{1:end-3},sort(varargin{end-2}),varargin{end-1:end});
            end
        end


because createByParams calls getContProbDynamicsBuilder and getContProbDynamicsBuilder calls createByParams.
Please make it so no method in LReachProblemDynamicsFactory  ever calls createByParams. This methos is only for external use.


9)Why do you use elseif here? Makes no sense

            if isDisturb
                linSys = handleObj(atStrCMat,btStrCMat,ptStrCMat, ...
                    ptStrCVec,ctStrCMat,qtStrCMat,qtStrCVec,x0Mat, ...
                    x0Vec,timeLimVec,relTol,absTol);
            elseif ~isDisturb
                linSys = handleObj(atStrCMat,btStrCMat,ptStrCMat, ...
                    ptStrCVec,x0Mat,x0Vec,timeLimVec,relTol,absTol);
            end

10) Please remove spaces and round brackets in if (~isDisturb && isBack) and the rest of the code in LReachProblemDynamicsFactory
11) Please make all lines in ReachContinuous/AReach/ReachDiscrete/LReachProblemDynamicsFactory shorter than 76 symbols where you can.
 getProDynamics violates this rule for instance. Also please apply smart-indentation to all the code (Ctr+A, Ctrl+I).

**Comment by ilya2015** [comment]
8)This is not a cyclic reference, becouse we call this method from

 gras.ellapx.**lreachplain**.probdyn.LReachProblemDynamicsFactory.createByParams, 
not for 
 gras.ellapx.**lreachuncert**.probdyn.LReachProblemDynamicsFactory.createByParams.
9-11) done

**Comment by pgagarinov** [comment]
11) is not fully done - please make lines shorter than 76 symbols, not just in the factory.
Then please merge and proceed to step 4.

**Comment by pgagarinov** [comment]
Implemented except for step 4 which is moved to a separate task #47
